cratetorrent/src/disk/io/piece.rs

```diff
@@ -88,15 +91,49 @@ impl Piece {
         torrent_piece_offset: u64,
         files: &[sync::RwLock<TorrentFile>],
     ) -> Result<(), WriteError> {
-        // convert the blocks to IO slices that the underlying
-        // systemcall can deal with
-        let mut blocks: Vec<_> = self
-            .blocks
-            .values()
-            .map(|b| IoVec::from_slice(&b))
-            .collect();
-        // the actual slice of blocks being worked on
-        let mut bufs = blocks.as_mut_slice();
+        cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                // convert the blocks to IO slices that the underlying
+                // pwritev syscall can deal with
+                let mut blocks: Vec<_> = self
+                    .blocks
+                    .values()
+                    .map(|b| IoVec::from_slice(&b))
+                    .collect();
+                // the actual slice of blocks being worked on
+                let mut bufs = blocks.as_mut_slice();
+            } else {
+                // (the same situation is in the `read` method below)
+                //
+                // We don't have pwritev The Convenient. Instead, we have two options:
+                //
+                // 1. Coalesce the slices to a continuous preallocated memory buffer and
+                //    write it at once.
+                // 2. Iterate through the slices and issue a separate write call for each.
+                //
+                // The first option implies preallocating a short-living buffer
+                // and destroying it shortly after. The second option
+                // implies issuing a big number of writes, and possibly disk I/O operations.
+                //
+                // We've chosen the first option. Piece size is commonly less than 1MB
+                // which is less than the page size, so the (re)allocation should be fast enough.
+                // Careful benchmarking, while necessary when the project gets to the optimization phase,
+                // is something I don't feel like doing just yet, and thus left as TODO
+                // and an exercise to the reader.
```

Comment:
Haha, funnily put but I agree.

Reply:
The comment is actually wrong: 1 MEGAbyte is bigger than page size (normally 4 KILObytes). I'll fix it.

Reply:
Ah, didn't catch that. That's correct. I have a suspicion that coalescing the buffers would still be faster. Opened for later https://github.com/mandreyel/cratetorrent/issues/98.